### ISSUE TITLE
Allow float targets for binary classification

### ISSUE THREAD

**Issue by freud14** [Lightning-AI/metrics#837]
In PyTorch, binary cross-entropy with and without logits expects floats as targets. However, if you try to use say the AUROC metric, you get an exception when passing the same targets. The exception comes from this line:

https://github.com/PyTorchLightning/metrics/blob/master/torchmetrics/utilities/checks.py#L33

Not sure about the impact of removing it or replacing it somewhere by `target = target.long()`.

**Comment by SkafteNicki** [comment]
Hi @freud14, thanks for this suggestions. I marked this issue to be part of a larger refactor of the classification package that we will start on soon.

**Comment by SkafteNicki** [comment]
Issue will be fixed by classification refactor: see this issue https://github.com/Lightning-AI/metrics/issues/1001 and this PR https://github.com/Lightning-AI/metrics/pull/1195 for all changes

Small recap: This issue ask that `target` can have dtype `float` for classification metrics. After the refactor, we have relaxed the constrain so most metric will work out of the box with `targets` that have dtype `float`. That said, we only guarantee that the metric works for `int/long` dtypes as we do not do any explicit conversion.  
```python
from torchmetrics.functional import binary_auroc
import torch
preds = torch.rand(10)
target = torch.randint(0, 2, (10,))

binary_auroc(preds, target.float()) # no error
```

Issue will be closed when https://github.com/Lightning-AI/metrics/pull/1195 is merged.
